File: scanner/module/dns/spf.go
package dns

import (
	"net"
	"strings"

	"github.com/bloom42/phaser/common/phaser"
	"github.com/bloom42/phaser/scanner/module"
)

type MissingOrInsufficientSPFRecord struct{}

func (MissingOrInsufficientSPFRecord) Name() string {
	return "dns/missing_or_insufficient_spf_record"
}

func (MissingOrInsufficientSPFRecord) Description() string {
	return "check if SPF record is insufficeient or missing"
}

func (MissingOrInsufficientSPFRecord) Author() string {
	return "Sylvain Kerkour <sylvain@kerkour.com>"
}

func (MissingOrInsufficientSPFRecord) Version() string {
	return "0.1.0"
}

type spfData struct {
	Domain    string   `json:"domain"`
	IsMissing bool     `json:"is_missing"`
	Records   []string `json:"records"`
}

func (MissingOrInsufficientSPFRecord) Run(scan *phaser.Scan, target *phaser.Target) (module.Result, []error) {
	errs := []error{}

	isSPFRecordMissing := true
	var ret module.Result
	records, err := net.LookupTXT(target.Host)

	if err != nil {
		errStr := err.Error()
		if strings.Contains(strings.ToLower(errStr), "no such host") == false {
			errs = append(errs, err)
			return ret, errs
		}
	}

	for _, record := range records {
		recordLower := strings.ToLower(record)
		if strings.Contains(recordLower, "v=spf1") {
			isSPFRecordMissing = false
			break
		}
	}

	// if "no such host"
	if records == nil {
		records = []string{}
	}

	if isSPFRecordMissing {
		data := spfData{
			Domain:    target.Host,
			Records:   records,
			IsMissing: isSPFRecordMissing,
		}
		ret = data
	}
	// if present, we found nothing so ret = nil
	return ret, errs
}

File: src/scanner/modules/domain/spf.rs
use crate::{
    scanner::{
        module,
        findings,
        Scan,
        Target,
        TargetKind,
    },
    error::PhaserError,
};
use std::process::{Command};

pub struct Spf{}

impl module::BaseModule for Spf {
    fn name(&self) -> String {
        return String::from("domain/spf");
    }

    fn description(&self) -> String {
        return String::from("check if DNS SPF record is missing or insufficeient");
    }

    fn author(&self) -> String {
        return String::from("Sylvain Kerkour <sylvain@kerkour.com>")
    }

    fn version(&self) -> String {
        return String::from("0.1.0");
    }
}

// TODO: only if root domain
impl module::HostModule for Spf {
    fn run(&self, _: &Scan, target: &Target) -> Result<findings::Data, PhaserError> {
        let mut errs = vec!();
        let mut ret = findings::Data::None;
        let mut is_spf_record_missing = true;

        if let TargetKind::Ip = target.kind {
            return Ok(findings::Data::None);
        };

        // first retrieve TXT records
        let dig_output = Command::new("dig")
            .arg("+short")
            .arg("TXT")
            .arg(&target.host)
            .output()?;
        let txt_output = String::from_utf8_lossy(&dig_output.stdout).to_string();

        let records: Vec<String> = txt_output.split('\n')
            .map(|record| record.trim().to_string())
            .filter(|record| !record.is_empty())
            .collect();

        let resolves = if records.len() != 0 { true } else { false };

        // for each record, check if DMARC data is present
        for record in &records {
            if record.to_lowercase().contains("v=spf1") {
                is_spf_record_missing = false;
                break;
            }
        }

        if is_spf_record_missing {
            ret = findings::Data::Spf(findings::domain::Spf{
                domain: target.host.clone(),
                records,
                resolves,
            });
        }

        return Ok(ret);
    }
}

#[cfg(test)]
mod tests {
    use crate::scanner::module::BaseModule;

    #[test]
    fn module_name() {
        let module = super::Spf{};
        assert_eq!("domain/spf", module.name());
    }
}
